Fix concurrency errors

weavejester opened the pull request:
I'm not sure if these are all the fixes needed, but this patch eliminates a lot of the issues I was having. The main problems were:

* `swap!` was used in a way that was not thread safe.
* The recursive SQL query did not correctly reference the working table, as in Que.
* There was no mechanism to lock the database connection while taking.

I have a feeling that this could be simplified further, but would require breaking changes, so I might write another library.

csummers commented:
@weavejester Thanks for these significant fixes!

I'm going to merge this into another branch and see if I can come up with a test case to try to reliably trigger the behavior you were experiencing.

weavejester commented:
I pushed a few thousand jobs onto the queue, then had a few dozen worker threads pull them off and attempt to process them. Sometimes the workers would delete them, other times they'd delete and then requeue, denoting a job that cannot be completed right now and should be retried. The tests were on a proprietary system, so I can't share them verbatim, but that was the gist of what I did.

On Friday I rewrote the queuing system entirely with some bespoke code. Rather than using atoms to maintain an in-memory lock state, I just hung onto the connection while the task was being processed. I discovered one more bug in my code that may exist in pgqueue as well. When an item is unlocked and deleted from a queue that is currently being taken from, the queue can suffer a phantom read - that is, the deleted row is read, because the locking take is seeing an older instance - and because the advisory locks exist independent of transactions, the deleted row is seen as an existing, unlocked row, resulting in a duplicate read.

To solve this issue, I simply added another query to check to make sure the item was still in the queue, after it has been taken.

tomjkidd commented:
@csummers Any chance this will become a released version?
